Support for Postgres Pooling

starrematte opened the pull request:
Added stable support for Pooling in Postgres. 
A few changes have been made in order to achieve a consistent code. 

- `deno.json` and `import_map.json` have been added for add support to old node libraries;

- in the `Connector` interface, for code design issues I had to put `_client `and `_connected` as optional due to inchoerence with the pool existence;

- `PostgresPoolOptions`  has been added and relies on the original options of Postgres Pool;

- `PostgresConnector` has been modified with substantial edits in the various exposed methods;

- `Database` has been modified with the underlying connector implementation of `_pool`.

eveningkid commented on import_map.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "imports": {
+    "https://dev.jspm.io/npm:@jspm/core@1/nodelibs/timers.js": "https://deno.land/std@0.159.0/node/timers.ts",
+    "https://dev.jspm.io/npm:@jspm/core@1/nodelibs/url.js": "https://deno.land/std@0.159.0/node/url.ts",
+    "https://dev.jspm.io/npm:@jspm/core@1/nodelibs/events.js": "https://deno.land/std@0.159.0/node/events.ts"
+  }
+}
```
I am sure this is a larger problem, but do you mind sharing what was causing this issue?

What parent lib?

Aiko-Suzuki replied:
The cause is https://github.com/eveningkid/denodb/blob/09eb6999a4849e1ef99a50a77dbd80a31639048b/deps.ts#L5 we can fix it by pointing to the branch instead of a commit.

https://github.com/Zhomart/dex/blob/update-deps/mod-dyn.ts

starrematte replied:
Yep sure, thanks!

eveningkid commented on lib/connectors/connector.ts:
```diff
@@ -16,19 +23,22 @@ export interface Connector {
 
   /** Client that maintains an external database connection. */
-  _client: ConnectorClient;
+  _client?: ConnectorClient | undefined;
```
nit: `x?: Y` automatically types it with `Y | undefined` as it's optional
```suggestion
  _client?: ConnectorClient;
```

eveningkid commented on lib/connectors/connector.ts:
```diff
@@ -16,19 +23,22 @@ export interface Connector {
 
   /** Is the client connected to an external instance. */
-  _connected: boolean;
+  _connected?: boolean | undefined;
+
+  /** Is the optional pool for making connections to an external instance. */
+  _pool?: ConnectionPool | undefined;
```
Same thing here, `| undefined` should be omitted :)

eveningkid commented on lib/connectors/connector.ts:
```diff
@@ -16,19 +23,22 @@ export interface Connector {
 
   /** Connect to an external database instance. */
-  _makeConnection(): void;
+  _makeConnection(): void | ConnectorClient | Promise<void> | Promise<ConnectorClient>;
```
We should keep this a `void` method

It makes a connection and shouldn't return anything

starrematte replied:
For the pool, we need it because of the underlying implementation

eveningkid commented on lib/connectors/postgres-connector.ts:
```diff
@@ -17,24 +17,53 @@ interface PostgresOptionsWithURI extends ConnectorOptions {
   uri: string;
 }
 
+interface PostgresPoolOptions extends ConnectorPoolOptions {
+  connection_params: PostgresOptionsWithConfig | PostgresOptionsWithURI,
+  size: number,
+  lazy: boolean,
+}
```
Let's keep everything on the same level and avoid `connection_params` here

```tsx
type PostgresClientOptions = PostgresOptionsWithConfig | PostgresOptionsWithURI;

interface PostgresPoolOptions extends ConnectorPoolOptions, PostgresClientOptions {
  size: number;
  lazy: boolean;
}

export type PostgresOptions = PostgresClientOptions | PostgresPoolOptions;
```

starrematte replied:
It should be like this, right? : 
``` typescript
type PostgresClientOptions = PostgresOptionsWithConfig & PostgresOptionsWithURI;
```

eveningkid commented on lib/connectors/postgres-connector.ts:
```diff
@@ -17,24 +17,53 @@ interface PostgresOptionsWithURI extends ConnectorOptions {
     if ("uri" in options) {
       this._client = new PostgresClient(options.uri);
-    } else {
+    } else if ("database" in options) {
```
Condition should be reversed with the other one below
1. `uri in options`
2. `size in options`
3. `else`

For 2), we can also use the new util `this._isPoolConnector()` mentioned in a comment below

eveningkid commented on lib/connectors/postgres-connector.ts:
```diff
@@ -43,25 +72,43 @@ export class PostgresConnector implements Connector {
   async _makeConnection() {
-    if (this._connected) {
-      return;
+    if (this._client) {
+      if (this._connected) {
+        return;
+      }
+
+      await this._client.connect();
+      this._connected = true;
+      return this._client;
+    } else {
+      return await this._pool!.connect();
```
Let's create two new methods:

```ts
_isPoolConnector() {
  return "size" in this._options;
}

_getClientOrPool() {
  return this._isPoolConnector() ? this.getPool() : this.getClient();
}

_makeConnection() {
  if (this._connected) {
    return;
  }

  await this._getClientOrPool().connect();
  this._connected = true;
}
```

eveningkid commented on lib/connectors/postgres-connector.ts:
```diff
@@ -43,25 +72,43 @@ export class PostgresConnector implements Connector {
   async ping() {
-    await this._makeConnection();
+    return await this.tryConnection(await this._makeConnection());
+  }
 
+  async tryConnection(client?: PostgresClient) {
     try {
       const [result] = (
-        await this._client.queryObject("SELECT 1 + 1 as result")
-      ).rows;
+        await client!.queryArray("SELECT 1 + 1 as result")
+      ).rows[0];
       return result === 2;
     } catch {
       return false;
```
Could we keep all the code inside `ping()`?

Now we'll need to call `this._getClientOrPool()` every time we need `this._client` or `._pool`

starrematte replied:
Yes, as we stick to the `this._getClientOrPool()` we can keep everything inside `ping()`

eveningkid commented on lib/connectors/postgres-connector.ts:
```diff
@@ -70,32 +117,35 @@ export class PostgresConnector implements Connector {
   async close() {
-    if (!this._connected) {
-      return;
+    if (this._client) {
+      if (!this._connected) {
+        return;
+      }
+      await this._client.end();
+    } else {
+      await this._pool?.end()!
     }
```
Same thing here, using `this._getClientOrPool().end()` would be great :)

eveningkid commented on lib/database.ts:
```diff
@@ -189,6 +189,11 @@ export class Database {
   }
 
+  /* Get the database pool if existent. */
+  getPool?() {
```
Can we drop the `?` here?

```suggestion
  getPool() {
```

eveningkid commented on lib/connectors/postgres-connector.ts:
```diff
@@ -17,24 +17,53 @@ interface PostgresOptionsWithURI extends ConnectorOptions {
+  _connected: boolean | undefined;
+
+  /** Create a PostgreSQL connection
+   *  @example 
+   * //Direct connection usage:
+   * const connection = new PostgresConnector({
+   *    host: '...',
+   *    username: 'user',
+   *    password: 'password',
+   *    database: 'airlines',
+   * });
+   * //Pool connection usage:
+   * const connection = new PostgresConnector({
+   *    connection_params: {
+   *      host: '...',
+   *      username: 'user',
+   *      password: 'password',
+   *      database: 'airlines',
+   *    },
+   *    size: 5,
+   *    lazy: false
+   * });
```
I know this comes from a good intention but I believe types are good hints to use the connector 

Types are always be up-to-date, we don't need to manually edit them each time

So let's keep your life easy here haha :D

starrematte replied:
Yeah you are right!

starrematte commented on lib/connectors/connector.ts:
```diff
@@ -1,11 +1,18 @@
+export interface ConnectorClient { }
+
+/** Default connector pool. */
+export interface ConnectionPool extends PostgresPool{ }
```
Yep, my bad, I had something in mind and left it there, thanks

starrematte commented:
new commit: https://github.com/eveningkid/denodb/pull/363/commits/3aeaec84bf20c61ccbf9ba531a3624205cc79e38

I fixed some stuff that I made and added yours. I did not test the other components involved with the `Connector`
changes.

If something seems strange, I forgot something or everything else, let's chat about it :)

eveningkid commented on lib/connectors/connector.ts:
```diff
@@ -16,19 +22,31 @@ export interface Connector {
-  _makeConnection(): void;
+  _makeConnection(): void | Promise<any>
+
+  /** Gets the client connected to the database */
+  getClient(): any
+
+  /** Gets the pool connected to the database */
+  getPool?(): any
```
```suggestion
  /** Gets the client connected to the database */
  getClient?(): ConnectorClient;

  /** Gets the pool connected to the database */
  getPool?(): ConnectorPool;
```

eveningkid commented on lib/connectors/connector.ts:
```diff
@@ -16,19 +22,31 @@ export interface Connector {
 
   /** Is the client connected to an external instance. */
-  _connected: boolean;
+  _connected?: boolean
```
This is always defined:

```suggestion
  _connected: boolean
```

eveningkid commented on lib/connectors/connector.ts:
```diff
@@ -16,19 +22,31 @@ export interface Connector {
 
   /** Client that maintains an external database connection. */
-  _client: ConnectorClient;
+  _client?: ConnectorClient;
```
Won't this cause type errors in other connectors?

Since it wasn't optional before, I assume that any code like:

```js
this.client.doSomething()
```

Would not work anymore, complaining that `this.client` could be not defined

We could keep `_client` as always defined in this file and override this interface just for `Postgres`

eveningkid reviewed (commented):
Another series of comments, thanks a lot for keeping up with them!

Also, please always run `deno fmt` at the root of your project before committing your files

It will keep all the syntax bits coherent with the rest of the codebase :)

starrematte commented on lib/connectors/postgres-connector.ts:
```diff
@@ -17,51 +17,86 @@ interface PostgresOptionsWithURI extends ConnectorOptions {
-      const [result] = (
-        await this._client.queryObject("SELECT 1 + 1 as result")
-      ).rows;
+      const connection = await this._makeConnection();
```
Forgot those while testing out😄
